server/sonar-web/src/main/js/app/components/nav/component/ComponentNavBranchesMenu.tsx

```diff
@@ -0,0 +1,216 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
+import { sortBy } from 'lodash';
+import ComponentNavBranchesMenuItem from './ComponentNavBranchesMenuItem';
+import { Branch, Component } from '../../../types';
+import { getBranches } from '../../../../api/branches';
+import { isShortLivingBranch, getBranchDisplayName } from '../../../../helpers/branches';
+import { translate } from '../../../../helpers/l10n';
+import { getProjectBranchUrl } from '../../../../helpers/urls';
+
+interface Props {
+  branch: Branch;
+  onClose: () => void;
+  project: Component;
+}
+
+interface State {
+  branches: Branch[];
+  loading: boolean;
+  query: string;
+  selected: string | null | undefined;
+}
+
+export default class ComponentNavBranchesMenu extends React.PureComponent<Props, State> {
+  private mounted: boolean;
+  private node: HTMLElement | null;
+
+  static contextTypes = {
+    router: PropTypes.object
+  };
+
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      branches: [],
+      loading: true,
+      query: '',
+      selected: null
+    };
+  }
+
+  componentDidMount() {
+    this.mounted = true;
+    this.fetchBranches();
+    window.addEventListener('click', this.handleClickOutside);
+  }
+
+  componentWillUnmount() {
+    this.mounted = false;
+    window.removeEventListener('click', this.handleClickOutside);
+  }
+
+  fetchBranches = () => {
+    this.setState({ loading: true });
+    getBranches(this.props.project.key).then(
+      (branches: Branch[]) => {
+        if (this.mounted) {
+          this.setState({ branches: this.sortBranches(branches), loading: false });
+        }
+      },
+      () => {
+        if (this.mounted) {
+          this.setState({ loading: false });
+        }
+      }
+    );
+  };
+
+  sortBranches = (branches: Branch[]): Branch[] =>
+    sortBy(
+      branches,
+      branch => !branch.isMain, // main branch first
+      branch => !isShortLivingBranch(branch), // then short-living branches
+      branch => getBranchDisplayName(branch) // then by name
+    );
+
+  handleClickOutside = (event: Event) => {
+    if (!this.node || !this.node.contains(event.target as HTMLElement)) {
+      this.props.onClose();
+    }
+  };
+
+  handleSearchChange = (event: React.SyntheticEvent<HTMLInputElement>) =>
+    this.setState({ query: event.currentTarget.value });
+
+  handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
```

Comment:
We can have both the search bar and the branch menu open at the same time and if we switch from one to another sometime the shortcuts stops working.